### ISSUE TITLE
Nrunner execution reporting -> 'Runnable' object is not subscriptable

### ISSUE THREAD

**Issue by marioaag** [avocado-framework/avocado#4500]
Running a basic implementation through the Python API hit the next issue:
```
Error running method "pre_tests" of plugin "fetchasset": 'Runnable' object is not subscriptable
JOB ID     : 2245c9b4d44bb802e55e4dfa89baed38db3bf48c
JOB LOG    : /Users/marioalvarado/avocado/job-results/job-2021-04-01T15.58-2245c9b/job.log

Avocado crashed: TypeError: 'Runnable' object is not subscriptable
Traceback (most recent call last):

  File "/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/core/job.py", line 608, in run_tests
    summary |= suite.run(self)

  File "/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/core/suite.py", line 279, in run
    return self.runner.run_suite(job, self)

  File "/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/plugins/runner.py", line 365, in run_suite
    test_factory[1]["base_logdir"] = job.logdir

TypeError: 'Runnable' object is not subscriptable

Please include the traceback info and command line used on your bug report
Report bugs visiting https://github.com/avocado-framework/avocado/issues/new
```

Code used:
```
    tests_reference = []
    test_suites = []
    for test_name in opts.tests_name:
        test_definition_path = os.path.join(opts.cfg_dir, 'testcases_definition',
                                            f'{test_name}.yml')
        test_values = utils.read_yaml(test_definition_path)
        test_reference = os.path.join(opts.cfg_dir, 'testcases', test_values['test_reference'])
        tests_reference.append(test_reference)

        # Avocado Framework Test Suites Generation
        test_suite = TestSuite(name=test_name, config={'run.test_parameters': [['host_ip', '300']]}, tests=[Runnable('avocado_instrumented', test_reference)])
        test_suites.append(test_suite)

    # Executing tests with Avocado Framework New Runner
    with Job(config={'run.test_runner': 'nrunner', 'core.show': ['app']}, test_suites=test_suites) as job:
        sys.exit(job.run())
```

Used the same parameters and ran using the standard avocado runner and it worked as expected:
```
    # Executing tests with Avocado Framework Actual Runner
    with Job.from_config(job_config={'run.references': tests_reference, 'run.test_parameters': [['host_ip', '300']]}) as job:
        sys.exit(job.run())
```

**Comment by clebergnu** [comment]
Hi @marioaag,

The first thing that comes to my eyes is that the runnable kind shuld be `avocado-instrumented` and not `avocado_instrumented`.  Please let me know if that fixes all the issues you're seeing.

Thanks.

**Comment by clebergnu** [comment]
Oh, one more thing.  You're not setting the `test.test_runner` configuration to [nrunner](https://avocado-framework.readthedocs.io/en/86.0/config/index.html#run-test-runner).  Without doing so, you'd be using the legacy `runner` implementation, which is not based on Runnables.

**Comment by willianrampazzo** [comment]
In addition to @clebergnu suggestion, also check if you are using the upstream version of Avocado. A similar problem was fixed here: https://github.com/avocado-framework/avocado/pull/4473

**Comment by marioaag** [comment]
@clebergnu Thanks for the comments, will do those changes...
@willianrampazzo is there a place with the instructions of how to use the upstream version? right now I'm using the official version 86.0

**Comment by marioaag** [comment]
@clebergnu I'm setting run.test_runner to nrunner, isn't it what you mean? I looked at the documentation and I didn't find anything related with test.test_runner

**Comment by clebergnu** [comment]
> @clebergnu I'm setting run.test_runner to nrunner, isn't it what you mean? I looked at the documentation and I didn't find anything related with test.test_runner

Yes, I only noticed after writing my comment.  I deleted it hoping that you'd not get notified about it.  So, in that regards (`test.test_runner`) this seems fine.

**Comment by willianrampazzo** [comment]
> @willianrampazzo is there a place with the instructions of how to use the upstream version? right now I'm using the official version 86.0

How have you installed Avocado on your system?

**Comment by willianrampazzo** [comment]
> > @willianrampazzo is there a place with the instructions of how to use the upstream version? right now I'm using the official version 86.0
> 
> How have you installed Avocado on your system?

If you installed using `pip`, remove the package with `pip uninstall avocado-framework` and install pointing to the upstream code with `pip install --user git+https://github.com/avocado-framework/avocado.git`.

**Comment by marioaag** [comment]
Did changes and installed the upstream version but the object not subscriptable issue persist:
```

JOB ID     : 33ec4ce4d96ea41ea98e8fdee8a30c5702e1bad4
JOB LOG    : /Users/marioalvarado/avocado/job-results/job-2021-04-06T09.17-33ec4ce/job.log

Avocado crashed: TypeError: 'Runnable' object is not subscriptable
Traceback (most recent call last):

  File "/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/core/job.py", line 620, in run_tests
    summary |= suite.run(self)

  File "/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/core/suite.py", line 311, in run
    return self.runner.run_suite(job, self)

  File "/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/plugins/runner.py", line 365, in run_suite
    test_factory[1]["base_logdir"] = job.logdir

TypeError: 'Runnable' object is not subscriptable

Please include the traceback info and command line used on your bug report
Report bugs visiting https://github.com/avocado-framework/avocado/issues/new
```

Looks like something like __setitem__ method definition is missing

**Comment by clebergnu** [comment]
@marioaag the traceback, such as in:

```
/Users/marioalvarado/Library/Python/3.7/lib/python/site-packages/avocado/plugins/runner.py
```

Shows that the `nrunner` is not really being used.  I wonder why.

**Comment by clebergnu** [comment]
@marioaag I can reproduce it with this code:

```python
import os
import sys

from avocado.core.job import Job
from avocado.core.suite import TestSuite
from avocado.core.nrunner import Runnable

tests_reference = []
test_suites = []
for test_name in ["X", "Y", "Z"]: #opts.tests_name:
    test_definition_path = os.path.join("/tmp", # opts.cfg_dir,
                                        'testcases_definition',
                                        f'{test_name}.yml')
    test_values = {'test_reference' : 'foo'} # utils.read_yaml(test_definition_path)
    test_reference = os.path.join("/tmp", #opts.cfg_dir,
                                  'testcases', test_values['test_reference'])
    tests_reference.append(test_reference)

    # Avocado Framework Test Suites Generation
    test_suite = TestSuite(name=test_name,
                           config={'run.test_parameters': [['host_ip', '300']]},
                           tests=[Runnable('avocado_instrumented', test_reference)])
    test_suites.append(test_suite)

# Executing tests with Avocado Framework New Runner
with Job(config={'run.test_runner': 'nrunner', 'core.show': ['app']},
         test_suites=test_suites) as job:
    sys.exit(job.run())
```

I'm looking at what's going on.

**Comment by clebergnu** [comment]
@marioaag I've found a problem in the merge of suites' and job's configuration.  To workaround it, please *also* include the `run.test_runner` configuration set to `nrunner` in the `TestSuite` configuration.  That is:

```python
    # Avocado Framework Test Suites Generation
    test_suite = TestSuite(name=test_name,
                           config={'run.test_runner': 'nrunner',
                                   'run.test_parameters': [['host_ip', '300']]},
                           tests=[Runnable('avocado_instrumented', test_reference)])

```

I'll open an issue or fix it right away.

**Comment by clebergnu** [comment]
@marioaag I was reviewing the API, and the proper way to use it (and avoid repeating configuration entries in the suite) is to provide the job config as an argument to the suite.  Example:

```python
import os
import sys

from avocado.core.job import Job
from avocado.core.suite import TestSuite
from avocado.core.nrunner import Runnable


JOB_CONFIG = {'run.test_runner': 'nrunner',
              'core.show': ['app']}


tests_reference = []
test_suites = []
for test_name in ["X", "Y", "Z"]: #opts.tests_name:
    test_definition_path = os.path.join("/tmp", # opts.cfg_dir,
                                        'testcases_definition',
                                        f'{test_name}.yml')
    test_values = {'test_reference' : 'foo'} # utils.read_yaml(test_definition_path)
    test_reference = os.path.join("/tmp", #opts.cfg_dir,
                                  'testcases', test_values['test_reference'])
    tests_reference.append(test_reference)

    # Avocado Framework Test Suites Generation
    test_suite = TestSuite(name=test_name,
                           config={'run.test_parameters': [['host_ip', '300']]},
                           tests=[Runnable('avocado-instrumented', test_reference)],
                           job_config=JOB_CONFIG)
    test_suites.append(test_suite)

# Executing tests with Avocado Framework New Runner
with Job(JOB_CONFIG,
         test_suites=test_suites) as job:
    sys.exit(job.run())
```

Please let me know if this works for you.

**Comment by marioaag** [comment]
@clebergnu, with this the issue gets resolved, but looks like it is ignoring use the nrunner, it is sequentially executing each one of the test:
```
JOB ID     : 0ada0605e618ac61af3defd8352b38697b802a19
JOB LOG    : /Users/marioalvarado/avocado/job-results/job-2021-04-07T14.43-0ada060/job.log
 (get_os_info-1/1) /Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_os_info: STARTED
 (get_os_info-1/1) /Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_os_info: PASS (5.21 s)
 (file_system_healt-1/1) /Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_file_system_healt: STARTED
 (file_system_healt-1/1) /Users/marioalvarado/git/iLabConfig/testcases/check_host_status.py:CheckHostStatus.test_file_system_healt: PASS (5.14 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /Users/marioalvarado/avocado/job-results/job-2021-04-07T14.43-0ada060/results.html
JOB TIME   : 12.61 s
```  

I tried run.test_runner: 'nrunner' in the TestSuite config and jobconfig also in the Job config without success.

```
    # Generating test paths
    tests_reference = []
    test_suites = []
    for test_name in opts.tests_name:
        test_definition_path = os.path.join(opts.cfg_dir, 'testcases_definition',
                                            f'{test_name}.yml')
        test_values = utils.read_yaml(test_definition_path)
        test_reference = os.path.join(opts.cfg_dir, 'testcases', test_values['test_reference'])
        tests_reference.append(test_reference)

        # Avocado Framework Test Suites Generation
        test_suite = TestSuite(name=test_name,
                               config={'run.test_runner': 'nrunner', 'run.test_parameters': [['host_ip', '300']]},
                               tests=[Runnable('avocado-instrumented', test_reference)],
                               job_config={'run.test_runner': 'nrunner', 'core.show': ['app']})
        test_suites.append(test_suite)

    # Executing tests with Avocado Framework New Runner
    with Job(config={'run.test_runner': 'nrunner', 'core.show': ['app']}, test_suites=test_suites) as job:
        sys.exit(job.run())
```

**Comment by willianrampazzo** [comment]
> @clebergnu, with this the issue gets resolved, but looks like it is ignoring use the nrunner, it is sequentially executing each one of the test:

Hi @marioaag, you are setting two test suites to run one test each. Tests inside a test suite run in parallel, but test suites run sequentially. Try adding more tests to your test suite and you will see them running in parallel.

**Comment by clebergnu** [comment]
Just adding a bit more context on top of @willianrampazzo explanation: each TestSuite is individually configurable, as you @marioaag know by now, even with regards to the test runner.  So, it'd be very complex at this point to process all test suites, each with potentially different configurations simultaneously.

But, this is planned, and once the nrunner is made the default runner implementation, it would be a tad simpler.

**Comment by marioaag** [comment]
Get it!! thanks so much for all the explanations, now it is running in parallel

The last thing that I'm missing for the moment is that the debug.log is empty when in the old runner I was getting the logging.

Any special consideration that I should take to get the logs printed int he debug.log file?

**Comment by clebergnu** [comment]
> Get it!! thanks so much for all the explanations, now it is running in parallel
> 
> The last thing that I'm missing for the moment is that the debug.log is empty when in the old runner I was getting the logging.
> 
> Any special consideration that I should take to get the logs printed int he debug.log file?

This is being fixed with #4507, we have hopes that it will be incorporated into Avocado 87.0 (to be made available next Monday).

**Comment by marioaag** [comment]
Nice! I will be waiting for it.
Whit all this said should I close this ticket? @clebergnu did you already open the ticket for the real issue found through all this?

**Comment by willianrampazzo** [comment]
> Nice! I will be waiting for it. Whit all this said should I close this ticket? @clebergnu did you already open the ticket for the real issue found through all this?

If I understood correctly, the real issue was in how the config was defined in the test. As it is working for you now, @marioaag, I'm closing this issue. Feel free to reopen it if the solution is not clear or open a new issue in case of unrelated to the original issue here.
